codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/OperationGenerator.java

```diff
@@ -141,116 +126,98 @@ public void run() {
-        }
-        Symbol decorateHandler = SymbolUtils.createValueSymbolBuilder(
-                "DecorateHandler", SmithyGoDependency.SMITHY_MIDDLEWARE).build();
-        Symbol newClientHandler = SymbolUtils.createValueSymbolBuilder(
-                "NewClientHandler", SmithyGoDependency.SMITHY_HTTP_TRANSPORT).build();
-        writer.write("handler := $T($T(options.HTTPClient), stack)", decorateHandler, newClientHandler);
-    }
-
-
     /**
      * Adds middleware to the operation middleware stack.
      */
-    private void populateOperationMiddlewareStack() {
-        // generate call to serde middleware helpers, these need to be generated first since other middleware will be
-        // added relative to them.
-        if (protocolGenerator != null) {
-            writer.write("$L(stack)", getOperationProtocolMiddlewareHelperName(operation));
-        }
-
-        // Populate middleware's from runtime client plugins
-        runtimeClientPlugins.forEach(runtimeClientPlugin -> {
-            if (!runtimeClientPlugin.matchesService(model, service)
-                    && !runtimeClientPlugin.matchesOperation(model, service, operation)) {
-                return;
-            }
-
-            if (!runtimeClientPlugin.registerMiddleware().isPresent()) {
-                return;
-            }
+    private void generateAddOperationMiddleware() {
+        Symbol stackSymbol = SymbolUtils.createPointableSymbolBuilder("Stack", SmithyGoDependency.SMITHY_MIDDLEWARE)
+                .build();
 
-            MiddlewareRegistrar middlewareRegistrar = runtimeClientPlugin.registerMiddleware().get();
-            Collection<Symbol> functionArguments = middlewareRegistrar.getFunctionArguments();
+        writer.openBlock("func $L(stack $P, options Options) (err error) {", "}",
+                getAddOperationMiddlewareFuncName(operationSymbol), stackSymbol,
+                () -> {
+                    generateOperationProtocolMiddlewareAdders();
+
+                    // Populate middleware's from runtime client plugins
+                    runtimeClientPlugins.forEach(runtimeClientPlugin -> {
+                        if (!runtimeClientPlugin.matchesService(model, service)
+                                && !runtimeClientPlugin.matchesOperation(model, service, operation)) {
+                            return;
+                        }
+
+                        if (!runtimeClientPlugin.registerMiddleware().isPresent()) {
+                            return;
+                        }
+
+                        MiddlewareRegistrar middlewareRegistrar = runtimeClientPlugin.registerMiddleware().get();
+                        Collection<Symbol> functionArguments = middlewareRegistrar.getFunctionArguments();
+
+                        // TODO these functions do not all return err like they should. This should be fixed.
+                        // TODO Must be fixed for all public functions.
```

Comment:
Do we have a GH issue on this?

Reply:
created https://github.com/awslabs/smithy-go/issues/211 to track this.